### PR TITLE
Fix for #8804 which can cause pages in CMS preview to be cached.

### PR DESCRIPTION
**Pull request by ichaber**
To solve the issue with cached responses after an edit in the CMS (described in issue #8804) augmentState in _HTTPCacheControlMiddleware_ set the HTTP headers to disable caching, if the request originates from the CMS preview.
The criteria to distinguish a request in the preview pane from any other request is the URL param **CMSPreview**. This will only be executed, if there is an active session.

### CODE REVIEW

File: src/Control/Middleware/HTTPCacheControlMiddleware.php [hunk 1]
```diff
@@ -796,11 +796,17 @@ protected function augmentState(HTTPRequest $request, HTTPResponse $response)
...
         } elseif ($request->getSession()->getAll()) {
-            // If sessions exist we assume that the responses should not be cached by CDNs / proxies as we are
-            // likely to be supplying information relevant to the current user only
 
-            // Don't force in case user code chooses to opt in to public caching
-            $this->privateCache();
+            if ($request->getVar("CMSPreview") === "1") {
+                //Disable caching when rendering the page in the CMS preview pane
+                //this is to prevent browsers from serving the local cache version after a page edit
+                $this->disableCache("true");
```
Comment (ScopeyNZ) [inline src/Control/Middleware/HTTPCacheControlMiddleware.php]:
```suggestion
                $this->disableCache(true);
```

**Comment by ScopeyNZ** [comment]
I'm not completely up to speed with our caching implementation but it seems the existing behaviour of setting cache to private should be correct?

**Comment by ScopeyNZ** [comment]
After some discussion I understand the reason for the change. `private` cache means the browser will assume it's cache-able where-as normally the CDN would cache it.

Still I'll defer to someone who has more context on the caching work in 4.2.

**Comment by maxime-rainville** [comment]
Is this fix meant to be against SilverStripe 5? If not, can you re-target to a `4` branch?

**Comment by sminnee** [comment]
Since this is a bug recommend `4.3`.

However, I'm not totally sold on the approach. This creates a CMS-specific dependency in the framework module. I think that this check should probably reside within `ContentController::init()` in the CMS module instead. You can access that via:

```php
HTTPCacheControlMiddleware::singleton()->disableCache(true)
```

**Comment by ichaber** [comment]
I'll create a new PR once I've changed the approach.

**Pull request closed without merging**